src/app_service/service/gif/service.nim

```diff
@@ -23,127 +29,196 @@ let TENOR_API_KEY_RESOLVED =
-
-    if(not response.error.isNil):
-      error "error getRecentGifs: ", errDescription = response.error.message
-
-    self.recents = map(response.result.getElems(), settingToGifDto)
-
-  except Exception as e:
-    error "error: ", procName="getRecentGifs", errName = e.name, errDesription = e.msg
-
-proc getFavoriteGifs(self: Service) =
-  try:
-    let response = backend.getFavoriteGifs()
-
-    if(not response.error.isNil):
-      error "error getFavoriteGifs: ", errDescription = response.error.message
-
-    self.favorites = map(response.result.getElems(), settingToGifDto)
+const SIGNAL_LOAD_RECENT_GIFS_STARTED* = "loadRecentGifsStarted"
+const SIGNAL_LOAD_RECENT_GIFS_FAILED* = "loadRecentGifsFailed"
+const SIGNAL_LOAD_RECENT_GIFS_DONE* = "loadRecentGifsDone"
 
-  except Exception as e:
-    error "error: ", procName="getFavoriteGifs", errName = e.name, errDesription = e.msg
+const SIGNAL_LOAD_FAVORITE_GIFS_STARTED* = "loadFavoriteGifsStarted"
+const SIGNAL_LOAD_FAVORITE_GIFS_FAILED* = "loadFavoriteGifsFailed"
+const SIGNAL_LOAD_FAVORITE_GIFS_DONE* = "loadFavoriteGifsDone"
 
-proc init*(self: Service) =
-  # set Tenor API Key
-  self.setTenorAPIKey()
-
-  # get recent and favorite gifs on the database
-  self.getRecentGifs()
-  self.getFavoriteGifs()
-
-proc tenorQuery(self: Service, path: string): seq[GifDto] =
-  try:
-    let response = backend.fetchGifs(path)
-    let doc = response.result.str.parseJson()
-
-    var items: seq[GifDto] = @[]
-    for json in doc["results"]:
-      items.add(tenorToGifDto(json))
-
-    return items
-  except:
-    return @[]
-
-proc search*(self: Service, query: string): seq[GifDto] =
-  return self.tenorQuery(fmt("search?q={encodeUrl(query)}"))
-
-proc getTrendings*(self: Service): seq[GifDto] =
-  return self.tenorQuery("trending?")
-
-proc getRecents*(self: Service): seq[GifDto] =
-  return self.recents
-
-proc addToRecents*(self: Service, gifDto: GifDto) =
-  let recents = self.getRecents()
-  var newRecents: seq[GifDto] = @[gifDto]
-  var idx = 0
-
-  while idx < MAX_RECENT - 1:
-    if idx >= recents.len:
-      break
-
-    if recents[idx].id == gifDto.id:
+type
+  GifsArgs* = ref object of Args
+    gifs*: seq[GifDto]
+    error*: string
+
+QtObject:
+  type
+    Service* = ref object of QObject
+      threadpool: ThreadPool
+      settingsService: settings_service.Service
+      favorites: seq[GifDto]
+      recents: seq[GifDto]
+      events: EventEmitter
+
+  proc delete*(self: Service) =
+    discard
+
+  proc newService*(settingsService: settings_service.Service, events: EventEmitter, threadpool: ThreadPool): Service =
+    result = Service()
+    result.QObject.setup
+    result.settingsService = settingsService
+    result.events = events
+    result.threadpool = threadpool
+    result.favorites = @[]
+    result.recents = @[]
+
+  proc setTenorAPIKey(self: Service) =
+    try:
+      let response = backend.setTenorAPIKey(TENOR_API_KEY_RESOLVED)
+      if(not response.error.isNil):
```

Comment:
I think you do not have to check err, because in case of error exception will be thrown.

Reply:
Actually haven't introduced this, but can remove it

Reply:
@endulab there are lot of such places, cause most likely we added check for an error in `core.nim`, later.